### PR TITLE
[Task] Message view models and list formatting

### PR DESCRIPTION
**Pull request by palatter**
## Purpose

* Formatting of the message list
* Split out message types

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/Message/MessageViewModel.swift [hunk 1]
```diff
@@ -6,13 +6,65 @@
...
+    let showDateHeader: Bool
+    let isConsecutive: Bool
+
+    init(message: ChatMessageInfoModel, showDateHeader: Bool, isConsecutive: Bool) {
+        self.message = message
+        self.showDateHeader = showDateHeader
+        self.isConsecutive = isConsecutive
+    }
+
+    var dateHeaderLabel: String {
```
Comment (palatter) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/Message/MessageViewModel.swift:19]: Will improve date logic with next PR

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift [hunk 2]
```diff
@@ -8,28 +8,57 @@ import Foundation
...
+            print("*Messages count: \(messageRepositoryManager.messages.count)") // for testing
+            for (index, message) in messageRepositoryManager.messages.enumerated() {
+                print("--*Messages: \(message.id) \(String(describing: message.content))") // for testing
+                messages.append(createViewModel(messages: messageRepositoryManager.messages, index: index))
             }
-//            self.messages = messageRepositoryManager.messages.toMessageViewModel
+        }
+    }
+
+    // Replace with factory
```
Comment (palatter) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift:38]: Will rework this to be optimized with next PR

**Review by jpeng-ms — DISMISSED** [review]
looks good

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift [hunk 3]
```diff
@@ -8,28 +8,56 @@ import Foundation
...
+    func createViewModel(messages: [ChatMessageInfoModel], index: Int) -> MessageViewModel {
+        let message = messages[index]
+        let lastMessageIndex = index == 0 ? 0 : index - 1
```
Comment (jpeng-ms) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift:41]: maybe we can use min() to get rid of this logic?

lastMessageIndex = min(0, index - 1)

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift [hunk 4]
```diff
@@ -8,28 +8,56 @@ import Foundation
...
+        let lastMessage = messages[lastMessageIndex]
+        let showDateHeader = index == 0 || message.createdOn.dayOfYear - lastMessage.createdOn.dayOfYear > 0
+        let isConsecutive = message.senderId == lastMessage.senderId
+
+        if messages[index].type == .text {
+            let isLocalUser = message.senderId == localUserId
+            let showUsername = !isLocalUser && lastMessage.senderId ?? "" != message.senderId ?? ""
+            let showTime = lastMessage.senderId ?? "" != message.senderId ?? ""
```
Comment (jpeng-ms) [inline AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift:49]: maybe we can make senderID optional and use guard here?

**Review by jimchou-dev — DISMISSED** [review]

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/Message/SystemMessageViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+class SystemMessageViewModel: MessageViewModel {
+    var content: String {
+        switch message.type {
+        case .participantsAdded:
+            return "\(participants) joined the chat"
+        case .participantsRemoved:
+            return "\(participants) left the chat"
+        default:
+            return "System Message"
```
[the remark this answers is not in the document]
Comment (palatter) [reply AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/Message/SystemMessageViewModel.swift:16]: There are a few system message types beyond joining/leaving. Will be covered in future work.

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift (hunk 4 above)
Comment (palatter) [reply AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift]: Will handle nil case at a later date

**Review by jimchou-dev — APPROVED** [review]

File: AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift (hunk 4 above)
Comment (palatter) [reply AzureCommunicationUI/sdk/AzureCommunicationUIChat/Sources/Presentation/SwiftUI/Chat/ChatViewComponents/MessageListViewModel.swift]: cleaned up the optionals though

**Pull request merged**